// @nativescript/schematics/src/nativescript-module/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const schematics_1 = require("@angular-devkit/schematics");
const utils_1 = require("../generate/utils");
const strings_1 = require("@angular-devkit/core/src/utils/strings");
const path_1 = require("path");
const angular_project_parser_1 = require("../angular-project-parser");
let nsext;
let projectSettings;
function default_1(options) {
    return schematics_1.chain([
        (tree) => {
            const nsconfigExtensions = utils_1.getNsConfigExtension(tree);
            nsext = options.nsext || nsconfigExtensions.ns;
            if (!nsext.startsWith('.')) {
                nsext = '.' + nsext;
            }
        },
        getProjectSettings,
        addNsFiles(options)
    ]);
}
exports.default = default_1;
const getProjectSettings = (tree, context) => {
    context.logger.info('Reading Project Settings');
    projectSettings = angular_project_parser_1.getAngularProjectSettings(tree);
};
const addNsFiles = (options) => (tree, context) => {
    context.logger.info('Adding Module {N} file');
    const templateOptions = {
        path: path_1.join(projectSettings.sourceRoot, 'app'),
        name: options.name,
        'if-flat': (s) => options.flat ? '' : s,
        classify: strings_1.classify,
        dasherize: strings_1.dasherize,
        nsext: nsext
    };
    const templateSource = schematics_1.apply(schematics_1.url('./_files'), [
        schematics_1.template(templateOptions)
    ]);
    return schematics_1.branchAndMerge(schematics_1.mergeWith(templateSource))(tree, context);
};
