// @nativescript/schematics/src/ng-new/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const schematics_1 = require("@angular-devkit/schematics");
function default_1(options) {
    return schematics_1.chain([
        (_tree) => {
            if (options.shared) {
                const sharedOptions = parseToSharedOptions(options);
                return schematics_1.schematic('shared', sharedOptions);
            }
            else {
                const applicationOptions = parseToApplicationOptions(options);
                return schematics_1.schematic('application', applicationOptions);
            }
        },
        () => {
            const npmInstallOptions = {
                workingDirectory: options.name
            };
            return schematics_1.schematic('npm-install', npmInstallOptions);
        }
    ]);
}
exports.default = default_1;
const parseToSharedOptions = (options) => {
    return {
        name: options.name,
        sourceDir: options.sourceDir || 'src',
        prefix: options.prefix,
        style: options.style,
        theme: options.theme,
        sample: options.sample,
    };
};
const parseToApplicationOptions = (options) => {
    return {
        name: options.name,
        prefix: options.prefix,
        sourceDir: options.sourceDir || 'app',
        style: options.style,
        theme: options.theme,
        webpack: options.webpack,
    };
};
